include/picongpu/particles/boundary/Apply.hpp

```diff
@@ -80,38 +73,32 @@ namespace picongpu
             template<typename T_Species>
             inline void apply(T_Species&& species, uint32_t currentStep)
             {
-                auto const numExchanges = NumberOfExchanges<simDim>::value;
                 auto const communicationMask = Environment<simDim>::get().GridController().getCommunicationMask();
-                for(uint32_t exchange = 1u; exchange < numExchanges; ++exchange)
+                for(auto exchange : getAllAxisAlignedExchanges())
                 {
-                    /* Here we are only interested in the positive and negative
-                     * directions for x, y, z axes and not the "diagonal" ones.
-                     * So skip other directions except left, right, top, bottom,
-                     * back, front
-                     */
-                    if(FRONT % exchange != 0)
-                        continue;
-
                     // If this is not an outer boundary, also skip
                     bool hasNeighbour = communicationMask.isSet(exchange);
                     if(hasNeighbour)
                         continue;
 
-                    // Transform exchange into axis
-                    uint32_t axis = 0;
-                    if(exchange >= BOTTOM && exchange <= TOP)
-                        axis = 1;
-                    if(exchange >= BACK)
-                        axis = 2;
+                    /* Here is the only place in the computational loop where we can call hooks from plugins:
+                     *    - we know those particles crossed the active boundary
+                     *    - but we didn't yet apply the boundary conditions
+                     *      that would normally modify or delete the particles
+                     *      (this is done just afterwards in this function)
+                     */
+                    detail::callPluginHooks(species, exchange);
```

Comment:
offline discussed: Particles that leave the volume to the diagonal neighbor volume will be accounted twice. In the old implementation, each of the 26 exchanges was called independently.

Reply:
In general you are right. I was making this change similarly to the new boundary conditions implementation, but it is not fully similar. However in this branch it's actually fine. Particles only leave when absorbing boundaries are used, and will be removed after crossing first such boundary and calling plugins once. With reflecting boundaries this will need to change, i got an idea how to do it

Reply:
Fixed now.